server/src/main/java/org/elasticsearch/common/time/DateMathParser.java

```diff
@@ -19,56 +19,31 @@
 
 package org.elasticsearch.common.time;
 
-import org.elasticsearch.ElasticsearchParseException;
+import org.joda.time.DateTimeZone;
 
-import java.time.DateTimeException;
-import java.time.DayOfWeek;
-import java.time.Instant;
-import java.time.LocalTime;
 import java.time.ZoneId;
-import java.time.ZoneOffset;
-import java.time.ZonedDateTime;
-import java.time.temporal.ChronoField;
-import java.time.temporal.TemporalAccessor;
-import java.time.temporal.TemporalAdjusters;
-import java.time.temporal.TemporalField;
-import java.time.temporal.TemporalQueries;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Objects;
 import java.util.function.LongSupplier;
 
 /**
- * A parser for date/time formatted text with optional date math.
- *
- * The format of the datetime is configurable, and unix timestamps can also be used. Datemath
- * is appended to a datetime with the following syntax:
- * <code>||[+-/](\d+)?[yMwdhHms]</code>.
+ * An abstraction over date math parsing to allow different implementation for joda and java time.
  */
-public class DateMathParser {
+public interface DateMathParser {
 
-    // base fields which should be used for default parsing, when we round up
-    private static final Map<TemporalField, Long> ROUND_UP_BASE_FIELDS = new HashMap<>(6);
-    {
-        ROUND_UP_BASE_FIELDS.put(ChronoField.MONTH_OF_YEAR, 1L);
-        ROUND_UP_BASE_FIELDS.put(ChronoField.DAY_OF_MONTH, 1L);
-        ROUND_UP_BASE_FIELDS.put(ChronoField.HOUR_OF_DAY, 23L);
-        ROUND_UP_BASE_FIELDS.put(ChronoField.MINUTE_OF_HOUR, 59L);
-        ROUND_UP_BASE_FIELDS.put(ChronoField.SECOND_OF_MINUTE, 59L);
-        ROUND_UP_BASE_FIELDS.put(ChronoField.MILLI_OF_SECOND, 999L);
+    /**
+     * Parse a date math expression without timzeone info and rounding down.
+     */
+    default long parse(String text, LongSupplier now) {
+        return parse(text, now, false, (ZoneId) null);
     }
 
-    private final DateFormatter formatter;
-    private final DateFormatter roundUpFormatter;
+    // Note: we take a callable here for the timestamp in order to be able to figure out
```

Comment:
We take a callable for "now", not for the timestamp. Right?

Reply:
Yes, these are only part of the parser. The parse method is what takes the actual date/time to parse alongside the math.

Reply:
Note that this comment here is old, it just looks new b/c it was moved around.